Framework/API/src/SpectrumInfo.cpp

```diff
@@ -0,0 +1,42 @@
+#include "MantidAPI/GeometryInfoFactory.h"
+#include "MantidAPI/GeometryInfo.h"
+#include "MantidAPI/SpectrumInfo.h"
+#include "MantidKernel/MultiThreaded.h"
+#include "MantidKernel/make_unique.h"
+
+namespace Mantid {
+namespace API {
+
+SpectrumInfo::SpectrumInfo(const MatrixWorkspace &workspace)
+    : m_factory{Kernel::make_unique<GeometryInfoFactory>(workspace)},
+      m_info(PARALLEL_GET_MAX_THREADS),
+      m_lastIndex(PARALLEL_GET_MAX_THREADS, -1) {}
+
+// Defined as default in source for forward declaration with std::unique_ptr.
+SpectrumInfo::SpectrumInfo(SpectrumInfo &&) = default;
+SpectrumInfo::~SpectrumInfo() = default;
+
+bool SpectrumInfo::isMonitor(const size_t index) const {
+  return getInfo(index).isMonitor();
+}
+
+bool SpectrumInfo::isMasked(const size_t index) const {
+  return getInfo(index).isMasked();
+}
+
+const GeometryInfo &SpectrumInfo::getInfo(const size_t index) const {
+  updateCachedInfo(index);
+  size_t thread = static_cast<size_t>(PARALLEL_THREAD_NUMBER);
+  return *m_info[thread];
+}
+
+void SpectrumInfo::updateCachedInfo(const size_t index) const {
```

Comment:
I am using the OpenMP thread ID as an index, just like `EventWorkspaceMRU`. The difference is that I keep only one entry per thread, unlike the MRU which keeps 50 and uses some really complicated boost map for that purpose.